File: packages/amp/py/spellguard_amp/server/channel.py
# SPDX-License-Identifier: Apache-2.0

"""
spellguard_amp.server.channel - Channel Management

Manage communication channels between agents.
"""

from __future__ import annotations

import time

from spellguard_amp.types import Channel

# In-memory channel storage
_channels: dict[str, Channel] = {}


def get_or_create_channel(agent1: str, agent2: str) -> Channel:
    """
    Get or create a channel between two agents.

    Args:
        agent1: First agent ID.
        agent2: Second agent ID.

    Returns:
        The channel (existing or newly created).
    """
    # Normalize channel ID (sorted to be consistent regardless of order)
    participants = tuple(sorted([agent1, agent2]))
    channel_id = f"channel_{participants[0]}_{participants[1]}"

    channel = _channels.get(channel_id)

    if channel is None:
        now = int(time.time() * 1000)
        channel = Channel(
            id=channel_id,
            participants=(participants[0], participants[1]),
            created_at=now,
            last_activity=now,
        )
        _channels[channel_id] = channel
        print(f"[AMP] Created channel: {channel_id}")

    return channel


def update_channel_activity(channel_id: str) -> None:
    """
    Update the last activity timestamp for a channel.

    Args:
        channel_id: Channel ID to update.
    """
    channel = _channels.get(channel_id)
    if channel is not None:
        channel.last_activity = int(time.time() * 1000)


def get_channel(channel_id: str) -> Channel | None:
    """
    Get channel by ID.

    Args:
        channel_id: Channel ID.

    Returns:
        Channel or None.
    """
    return _channels.get(channel_id)


def get_channel_stats() -> dict[str, int]:
    """
    Get statistics about channels.

    Returns:
        Dict with 'total', 'active', and 'stale' counts.
    """
    now = int(time.time() * 1000)
    stale_threshold = 24 * 60 * 60 * 1000  # 24 hours

    active = 0
    stale = 0

    for channel in _channels.values():
        if now - channel.last_activity > stale_threshold:
            stale += 1
        else:
            active += 1

    return {
        "total": len(_channels),
        "active": active,
        "stale": stale,
    }


def clear_channels() -> None:
    """Clear all channels (for testing)."""
    _channels.clear()

File: packages/amp/ts/src/server/channel.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * @spellguard/amp - Channel Management
 *
 * Manage communication channels between agents.
 */

import type { Channel } from '../types';

// In-memory channel storage
const channels = new Map<string, Channel>();

/**
 * Get or create a channel between two agents.
 *
 * @param agent1 - First agent ID
 * @param agent2 - Second agent ID
 * @returns The channel (existing or newly created)
 */
export function getOrCreateChannel(agent1: string, agent2: string): Channel {
  // Normalize channel ID (sorted to be consistent regardless of order)
  const participants = [agent1, agent2].sort() as [string, string];
  const channelId = `channel_${participants[0]}_${participants[1]}`;

  let channel = channels.get(channelId);

  if (!channel) {
    channel = {
      id: channelId,
      participants,
      createdAt: Date.now(),
      lastActivity: Date.now(),
    };
    channels.set(channelId, channel);
    console.log(`[AMP] Created channel: ${channelId}`);
  }

  return channel;
}

/**
 * Update the last activity timestamp for a channel.
 *
 * @param channelId - Channel ID to update
 */
export function updateChannelActivity(channelId: string): void {
  const channel = channels.get(channelId);
  if (channel) {
    channel.lastActivity = Date.now();
  }
}

/**
 * Get channel by ID.
 *
 * @param channelId - Channel ID
 * @returns Channel or undefined
 */
export function getChannel(channelId: string): Channel | undefined {
  return channels.get(channelId);
}

/**
 * Get statistics about channels.
 *
 * @returns Channel statistics
 */
export function getChannelStats(): {
  total: number;
  active: number;
  stale: number;
} {
  const now = Date.now();
  const staleThreshold = 24 * 60 * 60 * 1000; // 24 hours

  let active = 0;
  let stale = 0;

  for (const channel of channels.values()) {
    if (now - channel.lastActivity > staleThreshold) {
      stale++;
    } else {
      active++;
    }
  }

  return {
    total: channels.size,
    active,
    stale,
  };
}

/**
 * Clear all channels (for testing).
 */
export function clearChannels(): void {
  channels.clear();
}
